File: kinetics/reaction_classes/equilibrium_mass_action.py
from kinetics.reaction_classes.reaction_base_class import Reaction


class Uni_mass_action_eq(kinetics.Reaction):

    def __init__(self,
                 keq=None, kf=None,
                 a='', p='',
                 substrates=[], products=[]):

        super().__init__()

        self.parameter_names=[keq, kf]
        self.reaction_substrate_names = [a, p]
        self.substrates = substrates
        self.products = products

    def calculate_rate(self, substrates, parameters):

        # Substrates
        a = substrates[0]
        p = substrates[1]

        # Parameters
        keq = parameters[0]
        kf = parameters[1]

        fwd = kf*a
        rev = (kf/keq)*p

        rate = fwd-rev

        return rate

File: js_port/src/reactions/EquilibriumMassAction.js
const Reaction = require('../models/Reaction');

class Uni_mass_action_eq extends Reaction {
    constructor(keq, kf, a, p, substrates = [], products = []) {
        super();
        this.parameter_names = [keq, kf];
        this.reaction_substrate_names = [a, p];
        this.substrates = substrates;
        this.products = products;
    }

    calculateRate(substrates, parameters) {
        // Substrates
        const a = substrates[0];
        const p = substrates[1];

        // Parameters
        const keq = parameters[0];
        const kf = parameters[1];

        const catalytic_capacity = kf;

        if (a === 0 || p === 0) {
            return 0.0;
        }

        const thermodynamic_driving_force = 1 - (p / a / keq);

        return catalytic_capacity * thermodynamic_driving_force;
    }
}

module.exports = {
    Uni_mass_action_eq
};
